File: ocaml/lib/color.ml
type t =
  { r : float
  ; g : float
  ; b : float
  }

let from_rgb ~r ~g ~b = { r; g; b }
let white = { r = 1.; g = 1.; b = 1. }
let black = { r = 0.; g = 0.; b = 0. }
let random () = { r = Random.float 1.; g = Random.float 1.; b = Random.float 1. }
let add c c' = { r = c.r +. c'.r; g = c.g +. c'.g; b = c.b +. c'.b }
let combine c c' = { r = c.r *. c'.r; g = c.g *. c'.g; b = c.b *. c'.b }
let scale c f = { r = c.r *. f; g = c.g *. f; b = c.b *. f }
let lerp ~u ~v ~fraction = add (scale u (1.0 -. fraction)) (scale v fraction)

let clamp value ~lower ~upper =
  if value < lower then
    lower
  else if upper < value then
    upper
  else
    value


let bit8 v ~samples =
  let lower = 0.
  and upper = 0.999
  and scale = 1. /. Int.to_float samples in
  255.999 *. (scale *. v |> sqrt |> clamp ~lower ~upper) |> Float.to_int


let ppm { r; g; b } ~samples =
  let r = r |> bit8 ~samples in
  let g = g |> bit8 ~samples in
  let b = b |> bit8 ~samples in
  Printf.sprintf "%i %i %i" r g b


let to_string { r; g; b } = Printf.sprintf "{r=%f, g=%f, b= %f}" r g b

File: elixir_rt/lib/color.ex
defmodule Color do
  defstruct r: 0.0, g: 0.0, b: 0.0

  @doc ~S"""
  Parses the given `line` into a command.

  ## Examples
      iex> Color.lerp %Color{r: 0.0, g: 0.0, b: 0.0}, %Color{r: 1.0, g: 1.0, b: 1.0}, 0.5
      %Color{r: 0.5, g: 0.5, b: 0.5}

      iex> Color.ppm %Color{r: 1.0, g: 1.0, b: 1.0}, 1
      "255 255 255"

      iex> Color.ppm %Color{r: 0.0, g: 0.0, b: 0.0}, 1
      "0 0 0"
  """

  def black() do
    %Color{r: 0.0, g: 0.0, b: 0.0}
  end

  def white() do
    %Color{r: 1.0, g: 1.0, b: 1.0}
  end

  def add(%Color{r: r, g: g, b: b}, %Color{r: u, g: v, b: w}) do
    %Color{r: r + u, g: g + v, b: b + w}
  end

  def scale(%Color{r: r, g: g, b: b}, f) do
    %Color{r: f * r, g: f * g, b: f * b}
  end

  def lerp(%Color{} = u, %Color{} = v, f) do
    u |> scale(1.0 - f) |> add(v |> scale(f))
  end

  defp clamp(value, lower, upper) do
    cond do
      value < lower -> lower
      value > upper -> upper
      true -> value
    end
  end

  defp bit8(v, samples) do
    lower = 0.0
    upper = 0.999
    scale = 1.0 / samples

    (v * scale) |> :math.sqrt() |> clamp(lower, upper) |> Kernel.*(255.999) |> trunc()
  end

  def ppm(%Color{r: r, g: g, b: b}, samples) do
    "#{r |> bit8(samples)} #{g |> bit8(samples)} #{b |> bit8(samples)}"
  end
end
